### DISCUSSION TITLE
Injecting env variabales

### DISCUSSION THREAD

**Discussion by equinusocio** [pnpm/pnpm#5727] [Q&A]
Hello, is there a wat to inject env variables inside package.json like `NPM` does using the `config` key in package.json?

```json
"config": {
  "source": "src/index.css",
  "dist": "dist/open-elements.min.css"
},
"scripts": {
  "build": "postcss $npm_package_config_source -o $npm_package_config_dist"
}
```

While both npm and yarn support the `config` key, with pnpm i get an error:

```sh
test on main [!?] is 📦 v0.0.0 via ⬢ v16.15.0
➜ pnpm run build

> open-elements@0.0.0 build /Users/mattia/Repositories/open-elements
> run-s clean prepare:css build:css

 ERROR   ERROR  Unknown options: 'open-elements:source', 'open-elements:dist'
For help, run: pnpm help run
```

**Comment by zkochan** [comment]
This is currently not supported. You can use [cross-env](https://www.npmjs.com/package/cross-env).

**Reply by equinusocio** [reply]
Thanks for the answer, is there any plan for this? I don't want to add a deprecated lib just for this. I will stick to `npm` then. Thank you.

**Reply by zkochan** [reply]
I am not sure what is the problem with cross-env. It is a feature complete solution without known vulnerabilities.

Alternatively, you can run pnpm with the shell emulator. In that case setting env variables will work on all systems. Set [`shell-emulator=true`](https://pnpm.io/cli/run#shell-emulator).

Overall, I am not against supporting the config field in `package.json`, but it is the first time anyone asked about it.

**Reply by equinusocio** [reply]
cross-env is [deprecated since 2020](https://github.com/kentcdodds/cross-env/issues/257)

Of course there are other way to to add env variables, but the `config` is the built-in way without the need of extra tooling, deps or whatever, and i would like to stick on it on my projects.

Unfortunately there is not much documentation about it, just this
https://docs.npmjs.com/cli/v9/configuring-npm/package-json?v=true#config

**Reply by zkochan** [reply]
It literally says that it is finished. Not a single word about "deprecated" or "not maintained". It won't have new features but will release security fixes.

**Reply by zkochan** [reply]
Yarn supports loading from `.env` https://yarnpkg.com/features/scripting#environment-files

and Bun too: https://bun.sh/docs/runtime/env

We can add support for this too.

**Reply by ash14** [reply]
This seems relevant again with the release of pnpm 11 that doesn't permit environment variables to be declared in `.npmrc` any more.

The use case for us is to set variables that dependencies use for installation such as [CYPRESS_DOWNLOAD_MIRROR](https://docs.cypress.io/app/references/advanced-installation#Mirroring) and to avoid having the need to have everyone configure their own local env for this. This essentially rules out a cross-env approach.

Supporting `.env.pnpm` and/or an `envFile` config option sounds like a good solution for this.

**Comment by gapag** [comment]
What @equinusocio is asking for can be achieved as for now (v10.7.0) this way:

In your `pnpm-workspace.yaml`:

```yaml
name: clint
surname: eastwood
address:
  city: los angeles
```

In your `package.json` you can refer to it as `npm_config_<name of your var>`:

```json
{ 
// ...
  "scripts": {
    "use-vars": "echo ${npm_config_name} ${npm_config_surname}",
    "no-nested": "echo ${npm_config_name} lives in ${npm_config_address_city}",
  },
// ...
}
```

If you run `pnpm run use-vars`:
  * you get `clint eastwood`
 
If  you run `pnpm run no-nested`:
  * you get `clint lives in`
  * **Note the limitation - no nested properties allowed, or I could not guess the right syntax.**

The `npm` prefix makes me suspect that this is an unexpected feature rather than a planned one.

Is it so @zkochan ? I am mostly worried about how safe it is to use such feature.

This thread suggested me the solution: https://github.com/orgs/pnpm/discussions/6566

**Reply by zkochan** [reply]
We will probably stop populating the `npm_config_` env variables because npm doesn't want us to use their configuration system. Whether we will change it to `pnpm_config_` or simply remove, I don't know. But it will be a breaking change that will only happen in a new major version of pnpm.

**Reply by equinusocio** [reply]
@zkochan but is there an official doc about this behaviour (populating npm env vars)?

**Reply by zkochan** [reply]
No, what is the point of documenting it if we need to stop using `npm_config_`?
